### ISSUE TITLE
Tweening arbitrary values without involving DOM objects

### ISSUE THREAD

**Issue by mhelvens** [julianshapiro/velocity#455]
I'm currently using Velocity to animate my DOM, but I also need to animate my Three.js camera in 3D space. That's not a DOM element, but just a Javascript object with properties like `x`, `y` and `z`.

Yes, Velocity now has the `tween` property and the `progress` function, but they'll still only work if there is a DOM element in the mix. But using a dummy `$('<div>')` just seems silly. It'd be useful to get some explicit support for tweening arbitrary values in arbitrary objects.

**Comment by Rycochet** [comment]
As far as I'm aware there's no problems in creating a dummy element just for the animation without having to attach it to anything - it does seem a little bit of a workaround though ;-)

Often in this case I use the document itself, but often end up hooking rAF directly instead (ie, duplicating effort for document.hidden etc).

My other thought is that jQuery allows you to hook a javascript function/class instead (such as ```$(myfunc).triggerHandler("blah")```) - maybe hooking into that and abstracting in a different way might be useful...?

**Comment by mhelvens** [comment]
Yes, you can certainly trick the library into doing what you need. The question is: is that good API design?

**Comment by Rycochet** [comment]
Depends whether it's being used as designed - velocity is an animation engine, but only concerned with css (+scroll) on html and svg elements.

From what I've read looking through the source the only thing I can think of that is guaranteed to work is simply having a dummy element in order to leverage the various callbacks - but that dummy element would still need to implement some sort of interface somewhat similar to an actual dom element or there would need to be code in quite a few places checking if it's safe to read/write values - ie, about what we do now anyway (hopefully other people have better ideas when timezones coincide) ;-)

The alternative may be giving a slightly different path, or even a global progress callback. Maybe if someone took the velocity source, threw out everything not needed for animation frame based callbacks, then compared that to see what was actually needed? I've got a feeling that it could do almost everything needed in a couple of dozen lines of code - just that it's duplicating effort etc...

Probably needs @julianshapiro or someone else who knows the code almost as well as him to see if it's easily possible without any performance impact ;-)

**Comment by mhelvens** [comment]
If it's really only concerned with css (+scroll), then nothing should be done. Do one thing and do it well is usually a good way to design code. But it seems like a waste not to leverage the existing tweening / easing code for general use, like in Tween.js. (I'm now using both libraries in the same project, and I don't want to.)

And I suspect that Velocity has higher aims. Why else give us the `tween` property and `progress` function?

**Comment by Rycochet** [comment]
We asked for them :-P Hopefully by the time Julian gets onto this topic people will have gone through all the arguments and he'll think it's worthwhile adding :-P

**Comment by ydaniv** [comment]
@mhelvens true that.
IIRC this kind of refacor is pending on 2.0.

**Comment by ionflow** [comment]
@mhelvens would you mind posting an example of how you use velocity with three.js? I'm trying to animate the camera to position x,y,z but having some difficulty.

**Comment by mhelvens** [comment]
@ionflow: I actually never did, because of the problem I'm describing in this issue. I'm using Tween.js. Furthermore, the way in which I'm using Tween.js is a bit complicated, and therefore not really suitable for a simple example. A Google search for "tweening three.js camera" or somesuch will give you a number of useful results, however.

**Comment by ionflow** [comment]
Thanks @mhelvens, I was able to figure it out:
```javascript
    function tweenCamera(x,y,z){
        var startingX = camera.position.x;
        var startingY = camera.position.y;
        var startingZ = camera.position.z;
        var deltaX = x - startingX;
        var deltaY = y - startingY;
        var deltaZ = z - startingZ;

        webglCanvas.velocity({
            tween: [1,0]
        }, {
            progress: function(elements, complete, remaining, current, tweenValue) {
                camera.position.x = startingX + (deltaX * tweenValue);
                camera.position.y = startingY + (deltaY * tweenValue);
                camera.position.z = startingZ + (deltaZ * tweenValue);
            }
        });
    }
```
The velocity docs do touch on this, but it took me a bit to get it working properly: 
http://julian.com/research/velocity/#progress

**Comment by Rycochet** [comment]
Currently you can supply a forcefed string to Velocity as a ```"tween"``` property, then you can split that out etc.

In v2 (currently the develop branch) you can register new properties, have them animate, then simply ask for them in the progress, so they can each get different values and easings.

Documentation will catch up at some point, but it's specifically the ```"registerNormalization"``` action that will allow this (look at the ```src/Velocity/css/normalizations/display.ts``` file for ideas).
